File: noise/frame.go
package noise

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"
)

// writeFrame writes a length-prefixed frame to a connection.
func writeFrame(conn net.Conn, data []byte) error {
	length := uint32(len(data))
	if err := binary.Write(conn, binary.BigEndian, length); err != nil {
		return fmt.Errorf("write frame length: %w", err)
	}
	if _, err := conn.Write(data); err != nil {
		return fmt.Errorf("write frame data: %w", err)
	}
	return nil
}

// readFrame reads a length-prefixed frame from a connection.
func readFrame(conn net.Conn) ([]byte, error) {
	var length uint32
	if err := binary.Read(conn, binary.BigEndian, &length); err != nil {
		return nil, fmt.Errorf("read frame length: %w", err)
	}
	// Handshake messages are small — cap at 65536 bytes
	if length > 65536 {
		return nil, fmt.Errorf("frame too large: %d bytes", length)
	}
	data := make([]byte, length)
	if _, err := io.ReadFull(conn, data); err != nil {
		return nil, fmt.Errorf("read frame data: %w", err)
	}
	return data, nil
}

File: src/noise_chatbot/noise/frame.py
"""Length-prefixed frame I/O for Noise handshake messages.

<trl>
MODULE noise CONTAINS FUNCTION write_frame AND FUNCTION read_frame.
EACH RECORD frame SHALL CONTAIN INTEGER length AND DATA payload.
</trl>
"""

from __future__ import annotations

import socket

# <trl>DEFINE INTEGER handshake_frame_max_bytes AS 65536.</trl>
HANDSHAKE_FRAME_MAX_BYTES: int = 65_536


def write_frame(conn: socket.socket, data: bytes) -> None:
    """Write a 4-byte big-endian length prefix followed by ``data`` to ``conn``.

    <trl>
    FUNCTION write_frame SHALL WRITE INTEGER length AND DATA data TO RESOURCE conn.
    </trl>

    Go parity:
        ``noise/frame.go:writeFrame`` — ``binary.BigEndian`` uint32 prefix.
    """
    raise NotImplementedError("Phase C")


def read_frame(conn: socket.socket) -> bytes:
    """Read a length-prefixed frame; reject frames over 65 536 bytes.

    <trl>
    FUNCTION read_frame SHALL READ INTEGER length FROM RESOURCE conn
        THEN REJECT IF INTEGER length EXCEEDS 65536.
    FUNCTION read_frame SHALL READ DATA data FROM RESOURCE conn THEN RETURNS_TO SOURCE.
    </trl>

    Raises:
        ValueError: If the length prefix exceeds 65 536.
        OSError: On socket read failure (propagated).

    Go parity:
        ``noise/frame.go:readFrame`` — cap at 65 536.
    """
    raise NotImplementedError("Phase C")
